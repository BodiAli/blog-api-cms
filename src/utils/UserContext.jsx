import { createContext, useContext, useEffect, useState } from "react";

export const UserContext = createContext(null);

export function useUser() {
  return useContext(UserContext);
}

export function useIsUserLoggedIn() {
  const [signedIn, setSignedIn] = useState(false);
  const [loading, setLoading] = useState(true);
  useEffect(() => {
    const token = localStorage.getItem("token");

    async function fetchUser() {
      try {
        const res = await fetch(`${import.meta.env.VITE_SERVER_URL}/auth/validate`, {
          headers: {
            Authorization: token,
          },
        });

        if (res.ok) {
          setSignedIn(true);
          return;
        }
      } catch (error) {
        console.error(error.message);
        return;
      } finally {
        setLoading(false);
      }
    }
    fetchUser();
  }, []);

  return [signedIn, loading];
}
