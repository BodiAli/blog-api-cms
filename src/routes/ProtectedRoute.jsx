import { useState, useEffect } from "react";
import { Navigate } from "react-router";
import { toast } from "react-toastify";
import { UserContext } from "../utils/UserContext";
import Loader from "../components/Loader/Loader";

export default function ProtectedRoute({ children }) {
  const [user, setUser] = useState(null);
  const [redirect, setRedirect] = useState(false);
  const [loading, setLoading] = useState(true);

  const token = localStorage.getItem("token");

  useEffect(() => {
    if (!token) {
      toast.error("You need to sign in or sign up before continuing.");
      setLoading(false);
      setRedirect(true);
      return;
    }

    async function fetchUser() {
      try {
        const res = await fetch(`${import.meta.env.VITE_SERVER_URL}/auth/validate`, {
          headers: {
            Authorization: token,
          },
        });

        if (!res.ok) {
          if (res.status === 401) {
            toast.error("You need to sign in or sign up before continuing.");
            setRedirect(true);
            return;
          }
          throw new Error("Failed to fetch user");
        }

        const fetchedUser = await res.json();

        setUser(fetchedUser);
      } catch (error) {
        toast.error(error.message || "Error fetching user");
        setUser(null);
        setRedirect(true);
      } finally {
        setLoading(false);
      }
    }

    fetchUser();
  }, [token]);

  if (loading) return <Loader />;
  console.log("USER", user);

  if (redirect) {
    return <Navigate to="log-in" replace={true} />;
  }

  return <UserContext.Provider value={user}>{children}</UserContext.Provider>;
}
