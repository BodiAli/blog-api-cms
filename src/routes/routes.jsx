import App from "../App";
import ProtectedRoute from "./ProtectedRoute.jsx";
import Signup from "../pages/Signup.jsx";
import Login from "../pages/Login.jsx";
import NotFoundPage from "../pages/NotFoundPage.jsx";

const routes = [
  {
    path: "/",
    element: (
      <ProtectedRoute>
        <App />
      </ProtectedRoute>
    ),
    errorElement: <NotFoundPage />,
  },

  {
    path: "log-in",
    Component: Login,
  },

  {
    path: "sign-up",
    Component: Signup,
  },
];

export default routes;
