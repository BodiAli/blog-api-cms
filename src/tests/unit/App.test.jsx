import { render, screen, waitFor } from "@testing-library/react";
import { createMemoryRouter, RouterProvider } from "react-router";
import userEvent from "@testing-library/user-event";
import routes from "../../routes/routes";
import { UserContext } from "../../utils/UserContext";

const user = {
  firstName: "first name",
  lastName: "last name",
  Profile: { profileImgUrl: "imageUrl" },
};

vi.mock("../../routes/ProtectedRoute.jsx", async () => {
  return {
    default: vi.fn(({ children }) => {
      return <UserContext.Provider value={user}>{children}</UserContext.Provider>;
    }),
  };
});

vi.mock("date-fns", () => {
  return {
    formatDistanceToNow: vi.fn(() => "about 1 hour ago"),
  };
});

window.fetch = vi.fn(() => {
  return Promise.resolve({
    ok: true,
    status: 200,
    json: vi.fn(() => {
      return Promise.resolve({
        posts: [
          {
            id: 1,
            userId: 1,
            title: "Post 1",
            content: "Post content",
            published: true,
            likes: 5,
            createdAt: "2025-04-27T03:03:02.292Z",
            updatedAt: "2025-04-27T03:03:02.292Z",
            imgUrl: null,
            Topics: [
              { id: 1, name: "Topic 1" },
              { id: 2, name: "Topic 2" },
            ],
          },
        ],
        pages: 1,
      });
    }),
  });
});

describe("App component", () => {
  test("Should render header component with user data, and posts with expected management buttons and pagination", async () => {
    const router = createMemoryRouter(routes);

    let container;

    await waitFor(() => {
      const result = render(<RouterProvider router={router} />);

      container = result.container;
    });

    expect(container).toMatchSnapshot();
  });

  test("Should navigate to 'users/posts/1/edit' when card's edit link is clicked", async () => {
    const user = userEvent.setup();

    const router = createMemoryRouter(routes);

    await waitFor(() => {
      render(<RouterProvider router={router} />);
    });

    const editLink = screen.getByRole("link", { name: "Edit" });

    await user.click(editLink);

    expect(router.state.location.pathname).toBe("/users/posts/1/edit");
  });

  test("Should navigate to 'users/posts/1' when card's view post link icon is clicked", async () => {
    const user = userEvent.setup();

    const router = createMemoryRouter(routes);

    await waitFor(() => {
      render(<RouterProvider router={router} />);
    });

    const viewPostLink = screen.getByTestId("view-post");

    await user.click(viewPostLink);

    expect(router.state.location.pathname).toBe("/users/posts/1");
  });

  test("navbar should navigate to '/create-post' path when Create post link is clicked", async () => {
    const user = userEvent.setup();

    const router = createMemoryRouter(routes);
    render(<RouterProvider router={router} />);

    const createPostLink = screen.getByRole("link", { name: "Create post" });

    await user.click(createPostLink);

    expect(router.state.location.pathname).toBe("/create-post");
  });

  test("navbar should navigate to '/' path when Manage posts link is clicked", async () => {
    const user = userEvent.setup();

    const router = createMemoryRouter(routes, { initialEntries: ["/create-post"] });
    render(<RouterProvider router={router} />);

    const managePostsLink = screen.getByRole("link", { name: "Manage posts" });

    await user.click(managePostsLink);

    expect(router.state.location.pathname).toBe("/");
  });
});
