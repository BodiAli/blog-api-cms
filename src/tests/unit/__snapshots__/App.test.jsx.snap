// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App component > Should render user profile and posts with expected management buttons 1`] = `
<div>
  <header
    class="_header_63c2ea"
  >
    <div
      class="_websiteLogo_63c2ea"
    >
      <img
        alt="The Log Blog icon"
        src="data:image/svg+xml,%3c?xml%20version='1.0'%20encoding='iso-8859-1'?%3e%3c!--%20Uploaded%20to:%20SVG%20Repo,%20www.svgrepo.com,%20Generator:%20SVG%20Repo%20Mixer%20Tools%20--%3e%3csvg%20height='800px'%20width='800px'%20version='1.1'%20id='Layer_1'%20xmlns='http://www.w3.org/2000/svg'%20xmlns:xlink='http://www.w3.org/1999/xlink'%20viewBox='0%200%20512%20512'%20xml:space='preserve'%3e%3cpath%20style='fill:%23A5A5A5;'%20d='M512,467.478L512,467.478C512,492.066,492.066,512,467.478,512H44.522%20C19.933,512,0,492.066,0,467.478l0,0c0-24.588,19.933-44.522,44.522-44.522h422.957C492.066,422.957,512,442.89,512,467.478z'/%3e%3cpath%20style='fill:%23777777;'%20d='M169.739,478.609c0,4.61-3.738,8.348-8.348,8.348H72.348c-4.61,0-8.348-3.738-8.348-8.348%20s3.738-8.348,8.348-8.348h89.043C166.002,470.261,169.739,473.998,169.739,478.609z%20M439.652,470.261H272.696%20c-4.61,0-8.348,3.738-8.348,8.348s3.738,8.348,8.348,8.348h166.957c4.61,0,8.348-3.738,8.348-8.348S444.262,470.261,439.652,470.261%20z%20M234.852,470.261h-1.113c-4.61,0-8.348,3.738-8.348,8.348s3.738,8.348,8.348,8.348h1.113c4.61,0,8.348-3.738,8.348-8.348%20S239.462,470.261,234.852,470.261z%20M200.348,470.261h-1.113c-4.61,0-8.348,3.738-8.348,8.348s3.738,8.348,8.348,8.348h1.113%20c4.61,0,8.348-3.738,8.348-8.348S204.958,470.261,200.348,470.261z'/%3e%3cpath%20style='fill:%23CCCAC4;'%20d='M478.609,155.826v244.87c0,24.588-19.934,44.522-44.522,44.522H77.913%20c-24.588,0-44.522-19.934-44.522-44.522v-244.87c0-24.588,19.933-44.522,44.522-44.522h356.174%20C458.675,111.304,478.609,131.238,478.609,155.826z'/%3e%3cpath%20style='fill:%23F2EFE2;'%20d='M434.087,422.957H77.913c-12.295,0-22.261-9.966-22.261-22.261v-244.87%20c0-12.295,9.966-22.261,22.261-22.261h356.174c12.295,0,22.261,9.966,22.261,22.261v244.87%20C456.348,412.99,446.382,422.957,434.087,422.957z'/%3e%3cpath%20style='fill:%23BFBBA3;'%20d='M406.261,203.13H105.739c-9.22,0-16.696-7.475-16.696-16.696l0,0c0-9.22,7.475-16.696,16.696-16.696%20h300.522c9.22,0,16.696,7.475,16.696,16.696l0,0C422.957,195.655,415.481,203.13,406.261,203.13z%20M244.87,244.87L244.87,244.87%20c0-9.22-7.475-16.696-16.696-16.696H105.739c-9.22,0-16.696,7.475-16.696,16.696l0,0c0,9.22,7.475,16.696,16.696,16.696h122.435%20C237.395,261.565,244.87,254.09,244.87,244.87z%20M422.957,244.87L422.957,244.87c0-9.22-7.475-16.696-16.696-16.696H283.826%20c-9.22,0-16.696,7.475-16.696,16.696l0,0c0,9.22,7.475,16.696,16.696,16.696h122.435C415.482,261.565,422.957,254.09,422.957,244.87%20z'/%3e%3cpath%20style='fill:%23FFD880;'%20d='M406.261,384H105.739c-9.22,0-16.696-7.475-16.696-16.696v-66.783c0-9.22,7.475-16.696,16.696-16.696%20h300.522c9.22,0,16.696,7.475,16.696,16.696v66.783C422.957,376.525,415.482,384,406.261,384z'/%3e%3cpath%20style='fill:%23FC8059;'%20d='M139.688,374.23l-2.332,2.332l0.086,0.086l-19.098,19.098c-1.37,1.37-3.229,2.14-5.167,2.14l0,0%20c-1.938,0-3.797-0.77-5.167-2.14l-19.098-19.098l0.086-0.086l-2.332-2.332c-6.677-6.677-6.677-17.501,0-24.179l0,0%20c6.677-6.677,17.501-6.677,24.179,0l2.332,2.332l2.332-2.332c6.677-6.677,17.501-6.677,24.179,0l0,0%20C146.365,356.728,146.365,367.553,139.688,374.23z'/%3e%3cpath%20style='fill:%23D6A154;'%20d='M376.209,292.36l-22.261,29.682c-4.452,5.936-13.357,5.936-17.809,0l-22.261-29.682%20c-1.445-1.927-2.226-4.27-2.226-6.678V22.261C311.652,9.966,321.618,0,333.913,0h22.261c12.295,0,22.261,9.966,22.261,22.261%20v263.421C378.435,288.089,377.653,290.433,376.209,292.36z'/%3e%3cpath%20style='fill:%23B26932;'%20d='M370.087,300.522l-16.139,21.518c-4.452,5.936-13.357,5.936-17.809,0L320,300.522H370.087z'/%3e%3cpath%20style='fill:%23FFD880;'%20d='M311.652,256V22.261C311.652,9.966,321.618,0,333.913,0h22.261c12.295,0,22.261,9.966,22.261,22.261%20V256H311.652z'/%3e%3cpath%20style='fill:%23FCC159;'%20d='M345.043,256V0h11.13c12.295,0,22.261,9.966,22.261,22.261V256H345.043z'/%3e%3cpath%20style='fill:%23FC8059;'%20d='M378.435,22.261v22.261h-66.783V22.261C311.652,9.966,321.618,0,333.913,0h22.261%20C368.469,0,378.435,9.966,378.435,22.261z'/%3e%3c/svg%3e"
      />
      <span>
        The Log Blog
      </span>
    </div>
    <nav
      class="_nav_63c2ea"
    >
      <a
        aria-current="page"
        class="active"
        data-discover="true"
        href="/"
      >
        Manage posts
      </a>
      <a
        class=""
        data-discover="true"
        href="/create-post"
      >
        Create post
      </a>
      <a
        href="#"
      >
        Public
      </a>
    </nav>
    <div
      class="_profile_63c2ea"
    >
      <p>
        Welcome, 
        first name
         
        last name
      </p>
      <img
        alt="User profile picture"
        src="imageUrl"
      />
    </div>
  </header>
</div>
`;
